app/assets/javascripts/TortoiseJS/agent/view.coffee

```diff
@@ -369,3 +283,251 @@ class PatchDrawer
+
+    shapeTheta = Math.atan2(sy - y1, sx - x1) - Math.PI / 2
+
+    @view.ctx.translate(sx, sy)
+
+    if linkShape['direction-indicator'].rotate
+      @view.ctx.rotate(shapeTheta)
+    else
+      @view.ctx.rotate(Math.PI)
+
+    # Magic numbers c/o Esther -- JTT, JAB 4/13/15
+    scalingFactor = 4
+
+    onePixelThickness = 1 / scalingFactor / scalingFactor
+
+    # one pixel should == one patch (before scale) -- JTT 4/15/15
+    thickness = onePixelThickness * thickness / @view.onePixel
+
+    @view.ctx.scale(scalingFactor, scalingFactor)
+
+    @linkShapeDrawer.drawShape(@view.ctx, color, shapeName, thickness)
+
+    @view.ctx.restore()
+
+  drawLabel: (x, y, labelText, color) ->
+    @view.ctx.save()
+
+    @view.ctx.translate(x - 3*@view.onePixel, y + 3*@view.onePixel)
+    @view.ctx.scale(@view.onePixel, -@view.onePixel)
+    @view.ctx.textAlign = 'end'
+    @view.ctx.fillStyle = netlogoColorToCSS(color)
+    @view.ctx.fillText(labelText, 0, 0)
+
+    @view.ctx.restore()
+
+  draw: (link, turtles, canWrapX, canWrapY) ->
+    if not link['hidden?']
+      { end1, end2, color, thickness } = link
+      { xcor: e1x, ycor: e1y } = turtles[end1]
+      { xcor: e2x, ycor: e2y } = turtles[end2]
+
+      theta = @calculateLineAngle(e1x, e1y, e2x, e2y)
+
+      if thickness <= @view.onePixel
+        x1 = e1x
+        x2 = e2x
+        y1 = e1y
+        y2 = e2y
+        adjustedThickness = @view.onePixel
+      else
+        [x1, x2] = @adjustLinkEnds(link, e1x, e2x, Math.cos(theta), 180, 360)
+        [y1, y2] = @adjustLinkEnds(link, e1y, e2y, Math.sin(theta), 90,  270)
+        adjustedThickness = thickness
+
+      @view.ctx.strokeStyle = netlogoColorToCSS(color)
+      @view.ctx.lineWidth   = adjustedThickness
+
+      wrapX = @shouldWrapInDim(canWrapX, @view.patchWidth,  e1x, e2x)
+      wrapY = @shouldWrapInDim(canWrapY, @view.patchHeight, e1y, e2y)
+
+      @getWrappedLines(x1, y1, x2, y2, wrapX, wrapY).forEach(@_drawLinkLine(link, adjustedThickness))
+
+  _drawLinkLine: ({ color, size, heading, 'directed?': isDirected, shape: shapeName, label, 'label-color': labelColor }, thickness) => ({ x1, y1, x2, y2 }) =>
+
+    linkShape = @shapes[shapeName]
+    { curviness, lines } = linkShape
+
+    lines.forEach(
+      (line) =>
+
+        { 'x-offset': centerOffset, 'dash-pattern': dashPattern, 'is-visible': visible } = line
+
+        if visible
+
+          [xcomp, ycomp] = @calculateComps(x1, y1, x2, y2, size)
+          [xOff, yOff]   = @calculateSublineOffset(centerOffset, thickness, xcomp, ycomp)
+          offsetSubline  = @getOffsetSubline(x1, y1, x2, y2, xOff, yOff)
+          isCurved       = curviness > 0
+          isMiddleLine   = line is lines[1]
+          hasLabel       = label?
+
+          [midpointX, midpointY] = offsetSubline.midpoint()
+
+          if isCurved
+
+            [controlX,  controlY]  = @calculateControlPoint(midpointX, midpointY, curviness, xcomp, ycomp)
```

Comment:
If you always calculate control points, then all the logic below can be collapsed, since with `curviness == 0`,  `controlX == midpointX` and `controlY == midpointY`. `calculateControlPoint` is simple enough that I doubt the performance hit is worth the code complexity of the branches below.